### ISSUE TITLE
Parameter is Not All You Need: Starting from Non-Parametric Networks for 3D Point Cloud Analysis

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#24]
### Paper Link
https://arxiv.org/pdf/2303.08134.pdf
### Code Link
https://github.com/ZrrSkywalker/Point-NN

### Problems of Existing Work
To extract local geometries, adaptive point convolutions and graphlike message passing gain performance rises from scaling up the number of parameters, e.g., KPConv [61]’s 14.3M and PointMLP [39]’s 12.6M, is much larger than PointNet++’s 1.7M.

Can we achieve high 3D point cloud analysis performance using only these non-parametric components?

- farthest point sampling (FPS)
- k-Nearest Neighbors (k-NN)
- pooling operations

<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229394169-df028751-bc1f-4ab1-9f38-45cde62358d8.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229394666-5c8a7b15-00ac-4d79-b35c-02f3fe2c0bac.png">

## Methodology
<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229418377-e75bf6fe-0c1f-46ba-a991-2fa6dd0643b5.png">

### Raw-point Embedding 
Positional encoding in Transformer $\operatorname{PosE}\left(p_i\right)=\operatorname{Concat}\left(f_i^x, f_i^y, f_i^z\right) \in \mathbb{R}^{1 \times C_I}$

### FPS + k-NN 
- FPS is adopted to downsample the point number from $M$ to $M/2$
- k-NN, or ball query, is responsible for grouping k spatial neighbors for each center $c$ from the original $M$ points

### Local Geometry Aggregation

1. simply concatenate the neighbor feature $p_j$ with the center feature $p_c$ along the feature dimension: $f_{cj}=\operatorname{Concat}\left(f_c, f_j\right), \quad$ for $j \in \mathcal{N}_c$
2. weigh each $f_{cj}$ by the relative positional encoding:
$f_{c j}^w=\left(f_{c j}+\operatorname{Pos} \mathrm{E}\left(\Delta p_j\right)\right) \odot \operatorname{Pos} \mathrm{E}\left(\Delta p_j\right)$, where $\odot$ denotes element-wise multiplication.
3. both max and average pooling and obtain $\left\lbrace f_c^a, p_c\right\rbrace_{i=1}^{\frac{M}{2}}$:

$$
f_c^a=\operatorname{MaxP}(\lbrace f_{c j}^w\rbrace_{j \in \mathcal{N}_c})+\operatorname{AveP}(\lbrace f^w_{cj}\rbrace_{j \in \mathcal{N}_c})
$$

After 4 stages of non-parametric encoder, apply pooling operations to get a global representation $f_G$ with $C_G$ feature dimension of the input point cloud. 

### Point-Memory Bank
<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229460898-632b629b-4df5-4700-a352-a7705d233bfd.png">

**Memory Construction**

$$
\begin{aligned}
& F_{m e m}=\operatorname{Concat}\left(\left\lbrace\operatorname{EncNP}\left(P_n\right)\right\rbrace_{n=1}^N\right) \\
& T_{m e m}=\operatorname{Concat}\left(\left\lbrace\operatorname{OneHot}\left(t_n\right)\right\rbrace_{n=1}^N\right)
\end{aligned}
$$

**Similarity-based Prediction**

$$
S_{c o s}=\frac{f_G^t F_{m e m}^T}{\left\|f_G^t\right\| \cdot\left\|F_{m e m}\right\|} \in \mathbb{R}^{1 \times N}
$$

$$
\operatorname{logits}=\varphi\left(S_{\cos } T_{m e m}\right) \in \mathbb{R}^{1 \times K}
$$

where $\varphi(x)$ is an activation function. More similar feature memory of higher score contributes more to the ﬁnal classiﬁcation logits, and vice versa.

### 3D Object Detection
Point-NN can be utilized as a non-parametric classiﬁcation head for object detection. We sample the point cloud within each ground-truth 3D bounding box in the training set, and encode the object-wise features as the feature memory $F_{m e m}$.

### Point-PN
<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/229466432-288c0a6e-6fa4-4df5-bd4e-c59b5bf1a21c.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229466512-195fb88e-2071-4414-997d-be93bbdfcd2b.png">

We insert linear layers into the encoder’s every stage before and after the Geometry Extraction step. ‘C’ of ‘0+1’ denotes only appending one linear layer after the module, and ‘D’ of ‘1+1’ denotes inserting into both positions. The preceding layers pre-transform the point features to better reveal the local geometry, and the latter layers further parse the weighted features for high-level understanding. We observe that Point-PN of ‘1+2’ performs the best, and more learnable layers, ‘2+2’, would harm the performance.

### Segmentation 
<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229473427-da3eda60-b301-488e-b48b-c7e6c37c3308.png">

### Shape Classiﬁcation 
<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/229468641-80ba67b0-b4ec-4faa-a9a3-8a7d18435c2c.png">

### 3D Object Detection
<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/229468850-6ab87166-9ec8-47a2-bbc3-3fae1d6fbebf.png">

Only Average Recall is improved. 

## Notes
**Positional encoding** looks very helpful than encoding 3D coordinates only.
